Calculation refactor

mrdjohnson opened the pull request:
**no functionality changes**
added string names for basic calculations
creating classes for each calculation, this will allow testing and in-application checks to look cleaner,
continues to keeps all calculation related things in one spot
replacement of the basicOperator and the singleOperator lists for checking with the BasicCalculation and UnaryCalculation class types

mrdjohnson committed:
Calculator refactor
**no functionality changes**
added string names for basic calculations
creating classes for each calculation, this will allow testing and in-application checks to look cleaner,
continues to keeps all calculation related things in one spot
replacement of the basicOperator and the singleOperator lists for checking with the BasicCalculation and UnaryCalculation class types

wardellbagby commented on lib/app_component.dart:
```diff
@@ -49,7 +50,6 @@ class AppComponent {
 //  ];
 
   final RegExp nonNumberMatcher = new RegExp(r"[^0-9]");
```
Minor: You can do the same thing with a regex of \D

mrdjohnson replied:
Will look into it. Due to the changes in this PR this matcher will probably be factored out actually.

wardellbagby commented on lib/app_component.dart:
```diff
@@ -380,7 +379,7 @@ class AppComponent {
 
       //this fixes precision issues sqrt(2)*sqrt(2) = 2.000000000004
-      if (totalValue.toString().contains(".")) {
+      if (totalValue.toString().contains(decimal)) {
```
Bruh, you cheating....

mrdjohnson replied:
This was actually referenced [here](https://github.com/DocJava/Dartulator/pull/7). doing an "is double" check was not working even when Dart was set to strong typed mode. x.truncate() == x did not work either.  This is not an ideal fix and once a more promising fix comes along I will be moving over to it.

wardellbagby replied:
I want to take it a step deeper and get rid of the fact that you have to truncate at all. I know there was a 3rd party BigDecimal-esque class in Dart. Why couldn't you use that?

mrdjohnson replied:
Hmm, I remember finding a similar class when trying to get around this. The problem was that It was based on an older version of dart and I think there were some compatibility issues. Also I might have had to refactor the entire project to use num instead of Dart (not sure about that though). If you can find it, link the 3rd part class you are referring to so I can make sure we're talking about the same one.

wardellbagby commented on lib/util/calculation.dart:
```diff
@@ -0,0 +1,281 @@
+  String toString() => "";
+
+// for the arc(sine cosine tangent) functions to switch the answer to degrees if need be
+  static Function arcTrigWrapper(num arcTrigFunction(x)) {
```
Perhaps it's worth it to also create a specific class for the Trig functions? You could move this logic into their base class instead of here. It's not relevant for all calculation types, just the Trig ones, right?

mrdjohnson replied:
Yes! That! Good plan.

wardellbagby commented on lib/util/calculation.dart:
```diff
@@ -0,0 +1,281 @@
+  num call(num x, num y);
+}
+
+class BasicCalculation {}
```
If we have Binary and Unary, what is Basic for?

mrdjohnson replied:
When doing different checks in the app_component.dart file We needed to know which calculations were Basic or not to find out if a move was valid. More like this one are probably coming soon.

wardellbagby commented on lib/util/calculation.dart:
```diff
@@ -0,0 +1,281 @@
+}
+
+
+class MultiplyCalculation extends BinaryCalculation
```
Minor: Perhaps MultiplicationCalculation? 

Subtract -> Subtraction
Add -> Addition
Divide -> Division

Just flows better, in my opinion.

mrdjohnson replied:
Thought about that, still thinking about it. Would not hurt.

wardellbagby commented on lib/util/calculation.dart:
```diff
@@ -0,0 +1,281 @@
+}
+
+
+class BaseTenCalculation extends UnaryCalculation {
```
Seems like these two calculations should inherit from some base Base calculation. (Heh, base base...)

Perhaps a BaseNCalculation type that arbitrarily accepts a base for its constructor, and then have these two as subclasses since they're the most popular use of it. It looks like ExponentCalculator is almost this.

mrdjohnson replied:
Considering this, Since it is currently only for two classes I may leave as is

wardellbagby commented on lib/util/equation_processor.dart:
```diff
@@ -1,117 +1,93 @@
-  static process(List input) {
-    Queue<String> operators = new Queue();
+  static process(List equationList) {
+    Queue calculations = new Queue();
```
You should still make this typed. Just now, it should be typed for Calculations.

mrdjohnson replied:
No can do. The parenthesis are still strings not calculations.

wardellbagby replied:
Hmm.... Mixed object Queues aren't ideal. You know, as a Java programmer, I need typed _everything..._

As it stands, changing how parenthesis are done is out-of-scope, but it should be considered at some point in the future.

mrdjohnson replied:
As a fellow Java dev, I share that need, but as an aspiring Dart/web dev I am opening myself up to the new possibilities. 

Already considered! I am thinking of doing an EquationItem class, Calculation would implement this and the parentheses and numbers would fall under subset class types of this as well. Still trying to figure out a good design for all of this that would keep the code readable while also setting up good guidelines for future use.

mrdjohnson committed:
Calculator refactor
**no functionality changes**
adapted each class to _calculate vs call. this allows call to be overriden while keeping base functionality in the main class (example see trig classes)
continues to keeps all calculation related things in one spot
slight modification to the number matcher
